util/ros1msg/fieldskipper.go

```diff
@@ -0,0 +1,469 @@
+package ros1msg
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"math"
+
+	"github.com/wkalt/dp3/util/schema"
+)
+
+/*
+This file contains an implementation of the FieldSkipper interface for every
+type supported by ros1msg, including complex records and arrays. The point of
+this is to enable efficient extraction of values from a ROS1 message, without
+parsing data that we don't intend to extract.
+
+In particular, we do not extract variable-length arrays or fixed-length arrays
+with more than ten elements.
+
+The point of this distinction is to avoid keeping statistics on things like each
+element of a []uint8 byte array, while allowing us to keep statistics on more
+reasonable things like a [2]float64 of GPS coordinates.
+
+This is a fairly shaky heuristic, and the general strategy of indexing complex
+array types contains a lot of unknowns, because people use arrays for multiple
+purposes. Hopefully with time and experience we can improve it.
+*/
+
+////////////////////////////////////////////////////////////////////////////////
+
+var ErrShortRead = errors.New("short read")
+
+// FieldSkipper is an interface for Skipping and Emitting fields from a ROS1
+// message. The returned integers represent "bytes read". Skip will compute and
+// return the required number of bytes, while avoiding any actual parsing that
+// it can. Emit will actually parse the values and append them to the data
+// array.
+type FieldSkipper interface {
+	Skip(data []byte) (int, error)
+	Emit(data []byte, values *[]any) (int, error)
+}
+
+type Bool struct{}
```

Comment:
these types should be made private with some naming convention. They are capitalized because the lowercase names conflict with builtins.

Reply:
may not need the types, if we combine skip and emit into one parameterized function and type that